Convert codebase to module format

hellochar opened the pull request:
WIP at converting the codebase + build system to work well with modules. The diff is massive because we've reduced indentation for pretty much every source code file. Please review individual commits to better understand how things have changed (in particular, "import modules around the codebase"). 

TODO
- [x] codebase compiles with modules
- [x] play nice with build system
- [x] build standalone
- [x] update tests

Fixes #3052

hellochar committed:
delete all Plottable namespace declarations

also make tsconfig.json blob based

hellochar committed:
reformat codebase (for whitespace)

other things may have been reformatted by IntelliJ

hellochar committed:
maintain Plottable namespace structure

folders (except for core/) now have index.ts files whose exports
correspond to the namespace for that folder. src/index.ts further
nests all these folder-level exports within a single export which
represents the Plottable namespace. Also add a namespacingTest that just
references various declarations to ensure it still works

hellochar committed:
import modules around the codebase

modify tsconfig to allow non-relative imports

fix all errors emitted by tsc 2.1.x

hellochar committed:
introduce npm command to run tsc

CalvinFernandez commented on src/components/plotGroup.ts:
```diff
@@ -1,8 +1,18 @@
-    this.components().forEach((plot: Plot) => {
+    this.components().forEach((plotComponent: Component) => {
+      // we know it's a Plot since .append() throws a runtime error otherwise
+      const plot: Plot = <Plot> <any> plotComponent;
```
`<Plot> <any>` ?

CalvinFernandez commented on src/axes/categoryAxis.ts:
```diff
@@ -126,7 +134,7 @@ export class Category extends Axis<string> {
    * @return {DownsampleInfo} an object holding the resultant domain and new stepWidth.
    */
-  public getDownsampleInfo(scale = <Scales.Category> this._scale): DownsampleInfo {
+  public getDownsampleInfo(scale: Scales.Category = <Scales.Category> this._scale): DownsampleInfo {
```
I think you can remove one of these `Scales.Category` notations

CalvinFernandez commented on namespacingTests.ts:
```diff
@@ -42,7 +42,7 @@ type TestInterfaces = Plottable.Axes.DownsampleInfo
   | Plottable.Plots.AccessorScaleBinding<any, any>
   | Plottable.Scales.TickGenerators.TickGenerator<any>
-  | Plottable.Scales.PaddingExceptionsProvider
+  | Plottable.Scales.PaddingExceptionsProvider<any>
```
why is this generic now?

CalvinFernandez commented on src/index.ts:
```diff
@@ -0,0 +1,79 @@
+import * as Animators from "./animators";
```
whats up with the spacing in this file?

hellochar replied:
The whitespace reflects the mapping between file hierarchy to namespace hierarchy, especially because the exports vary later on.

Oh I just realized I forgot to be consistent with these first three... will change

CalvinFernandez replied:
can we not do that -- it makes it hard to maintain

CalvinFernandez commented on src/plots/index.ts:
```diff
@@ -0,0 +1,55 @@
+  scale?: TransformableScale<D, R>;
+}
+
+export namespace Animator {
```
We're removing namespaces. This should go in its own plots/animator/index.ts

CalvinFernandez commented on src/plots/index.ts:
```diff
@@ -0,0 +1,55 @@
+}
+
+export namespace Animator {
+  export var MAIN = "main";
```
While you're at it `export const`

CalvinFernandez commented on src/scales/index.ts:
```diff
@@ -0,0 +1,91 @@
+ * If one end of the domain is set to an excepted value as a result of autoDomain()-ing,
+ * that end of the domain will not be padded.
+ *
+ * @param {QuantitativeScale} scale
```
While you're here: `QuantitativeScale<D>`

adidahiya replied:
echoing a comment in another PR: why do we maintain types in JSDoc?

hellochar replied:
Not actually sure... might not even be needed

CalvinFernandez replied:
Yea i don't think its needed 

http://typedoc.org/guides/doccomments/ 

let's abolish ?

CalvinFernandez reviewed (commented):
looking good

adidahiya commented:
Try adding ?w=1 at the end of the diff URL. The diff becomes manageable https://github.com/palantir/plottable/pull/3184/files?w=1

CalvinFernandez commented on src/index.ts:
```diff
@@ -0,0 +1,79 @@
+  Drawers,
+
+  Interactions,
+
```
the spacing here is a nice thought but its just too much to maintain

hellochar committed:
remove non-relative paths so grunt compile works

hellochar committed:
test files compile

also fix various mised export in src/

grunt's ts:test ignores outDir

hellochar committed:
delete testReference, replace with imports

hellochar committed:
remove circular deps from src modules

so bundling with e.g. webpack will build a runnable file

hellochar committed:
use webpack to bundle + umd plottable.js; update .d.ts

add two webpack configs to build plottable.js and plottable-npm.js,
respectively. retains previous files' behaviors (aka no breaking
change).

.d.ts files now point to the emitted .d.ts files in /build

hellochar committed:
fix plottable.d.ts compile error

hellochar committed:
build tests.js from webpack

hellochar committed:
remove unused jquery.simulate and @types packages

hellochar committed:
make tests pass

hellochar committed:
adapt the gruntfile to work with modules

* delegate to npm for building/bundling
* disable coverage for the time being

hellochar committed:
use node 5.6.0 in circle; remove whitespace

hellochar committed:
remove unused plottable.js in test files

hellochar committed:
Merge branch 'develop' into xzhang/modules

Conflicts:
	Gruntfile.js
	circle.yml
	package.json
	plottable-npm.d.ts
	plottable-npm.js
	plottable.d.ts
	plottable.js
	plottable.zip
	src/components/legend.ts
	src/utils/mathUtils.ts

hellochar committed:
ensure ts compilation includes all src files

hellochar committed:
move webpack demo config to top-level

CalvinFernandez commented on webpack.config.demo.js:
```diff
@@ -0,0 +1,9 @@
+var path = require("path");
```
can we move the webpack configs into `/config/` ?

CalvinFernandez commented on test/coverage.html:
```diff
@@ -23,6 +23,9 @@
     <script src="../node_modules/sinon/lib/sinon/spy.js"></script>
     <!--  --> 
     <script src="blanket_mocha.js"></script>
+    <!-- hack-hack this Plottable file isn't ever used since tests.js bundles one
```
HACKHACK, so we can still grep

CalvinFernandez commented on Gruntfile.js:
```diff
@@ -72,20 +71,12 @@ module.exports = function(grunt) {
 
   var concatConfig = {
     header: {
-      src: ["license_header.txt", "plottable.js", "node_modules/svg-typewriter/svg-typewriter.js"],
```
hopefully webpack pulls in svg-typewriter as part of the bundle now?

CalvinFernandez replied:
actually never mind i see what's happening now

CalvinFernandez commented on test/tsconfig.json:
```diff
@@ -0,0 +1,14 @@
+    "outDir": "../build",
+    "noImplicitAny": true,
+    "removeComments": false,
+    "sourceMap": false,
```
can we turn this on

CalvinFernandez reviewed (commented):
its happening!

hellochar committed:
upgrade version of safari being tested

hellochar committed:
Merge branch 'develop' into xzhang/modules

Conflicts:
	Gruntfile.js
	plottable-npm.d.ts
	plottable-npm.js
	plottable.d.ts
	plottable.js
	src/interactions/panZoomInteraction.ts
	src/plots/barPlot.ts
	src/plots/stackedAreaPlot.ts
	src/plots/stackedBarPlot.ts
	src/utils/stackingUtils.ts

CalvinFernandez reviewed (approved):
lets go

hellochar committed:
add build/ to npm and bower packages

CalvinFernandez commented:
🚢

hellochar committed:
plottable.d.ts doesn't depend on plottable-npm.d.ts

hellochar committed:
Merge branch 'develop' into xzhang/modules

Conflicts:
	plottable.js
	plottable.min.js
	plottable.zip
